analyzers/src/SonarAnalyzer.CSharp/Rules/ParameterTypeShouldMatchRouteTypeConstraint.cs

```diff
@@ -0,0 +1,171 @@
+            {
+                return;
+            }
+
+            cc.RegisterNodeAction(c =>
+            {
+                var node = (ClassDeclarationSyntax)c.Node;
+
+                foreach (var property in GetPropertyTypeMismatches(node, c.SemanticModel))
+                {
+                    var additionalLocations = property.RouteParamLocation is not null
+                        ? new[] { property.RouteParamLocation }
+                        : Array.Empty<Location>();
+                    var message = property.RouteParamLocation is not null
+                        ? string.Format(MessageWithSecondaryLocationFormat, property.Type.GetName())
+                        : string.Format(MessageNoSecondaryLocationFormat, property.Type.GetName(), property.ConstraintType, property.Route);
+
+                    c.ReportIssue(Diagnostic.Create(Rule,
+                        property.Type.GetLocation(),
+                        additionalLocations,
+                        additionalLocations.ToProperties(string.Format(SecondaryMessageFormat, property.ConstraintType)),
+                        message));
+                }
+            },
+            SyntaxKind.ClassDeclaration);
+        });
+
+    private static IList<PropertyTypeMismatch> GetPropertyTypeMismatches(ClassDeclarationSyntax classDeclaration, SemanticModel semanticModel)
+    {
+        var routeParameters = GetRouteParametersWithValidConstraint(classDeclaration);
+
+        if (routeParameters.Count == 0)
+        {
+            return Array.Empty<PropertyTypeMismatch>();
+        }
+
+        var mismatches = new List<PropertyTypeMismatch>();
+        var classSymbol = semanticModel.GetDeclaredSymbol(classDeclaration);
+        var properties = classSymbol.GetMembers()
+            .Where(member => member.Kind is SymbolKind.Property && member.HasAttribute(KnownType.Microsoft_AspNetCore_Components_ParameterAttribute) && routeParameters.ContainsKey(member.Name));
+
+        foreach (var property in properties)
+        {
+            foreach (var routeParameter in routeParameters[property.Name])
+            {
+                if (TypeMatchConstraint(property.GetSymbolType(), routeParameter.Constraint))
+                {
+                    continue;
+                }
+
+                foreach (var propertySyntax in property.DeclaringSyntaxReferences.Where(r => r.GetSyntax() is PropertyDeclarationSyntax).Select(r => (PropertyDeclarationSyntax)r.GetSyntax()))
+                {
+                    mismatches.Add(new PropertyTypeMismatch(propertySyntax.Type, routeParameter.Constraint, routeParameter.FromRoute, routeParameter.RouteParamLocation));
+                }
+            }
+        }
+
+        return mismatches;
+    }
+
+    private static bool TypeMatchConstraint(ISymbol type, string constraintType) =>
+        ConstraintMapping.ContainsKey(constraintType) && ConstraintMapping[constraintType] == type.Name;
```

Comment:
What about implicit conversions:
* int -> int? (Nullable conversion)
* int -> object (boxing conversion)
* int -> IComparable<int> (boxing conversion)
* int -> long (Widening conversion)

Basically any of these:
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/conversions#102-implicit-conversions

Plus according to the [docs](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-7.0#route-constraints)

> Route constraints also work with [optional parameters](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-7.0#route-parameters). In the following example, Id is required, but Option is an optional boolean route parameter.

It seems an explicit conversion from bool? to bool is also supported.

There is a [CSharpExtensions.ClassifyConversion](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.csharpextensions.classifyconversion?view=roslyn-dotnet-4.6.0) which gives you [Conversion.IsImplicit](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.csharp.conversion.isimplicit?view=roslyn-dotnet-4.6.0#microsoft-codeanalysis-csharp-conversion-isimplicit)

Reply:
I added more tests for implicit, boxing, widening, and also enumeration conversion.

> It seems an explicit conversion from bool? to bool is also supported.

Optional parameters are not nullable, there are about the route parameter being optional when resolving the route. If the optional route parameter is missing, the default value of the property type is used.